### ISSUE TITLE
false error on returned class when type hinded with Self

### ISSUE THREAD

**Issue by GalGof** [python/mypy#16317]
**Bug Report**

wrongly produce error when returned objects is same class instance as Self class

**To Reproduce**
```from typing import Self


class AAA:
    def copy(self)-> Self:
        return AAA()```

**Expected Behavior**

no issues

**Actual Behavior**

(error: Incompatible return value type (got "AAA", expected "Self")  [return-value])


**Your Environment**
mypy Version: 1.6.1
python 3.11

**Comment by erictraut** [comment]
Mypy's behavior is correct here. The `Self` type represents the type of the class _or a subclass_. If you were to subclass `AAA` and call the `copy` method, it would not properly return a copy of that subclass instance.

You can fix this in your code by doing one of two things:
1. Change the return type annotation to `AAA` to indicate that `copy` returns an instance of `AAA` regardless of whether it's called on a subclass instance.

```python
class AAA:
    def copy(self) -> AAA:
        return AAA()
```

2. Change your implementation of `copy` to return a value that is compatible with `Self`:

```python
class AAA:
    def copy(self) -> Self:
        return type(self)()
```

Either of these will type check without error in mypy.
